[net-diag] add TLV for MLE counters

abtink opened the pull request:
This commit introduces Network Diagnostic TLV for MLE counters.
It updates `NetworkDiagnostic::Server` to provide this TLV when
queried and provide support for retrieving on `Client`.

----

This PR contains the commit from https://github.com/openthread/openthread/pull/8866

Related to [SPEC-1186](https://threadgroup.atlassian.net/browse/SPEC-1186).

Vyrastas commented on src/cli/cli.cpp:
```diff
@@ -5798,6 +5798,122 @@ template <> otError Interpreter::Process<Cmd("meshdiag")>(Arg aArgs[])
+    /**
+     * @cli meshdiag childip6
+     * @code
+     * meshdiag childrenip6addr 0xdc00
```
childrenip6addr --> childip6

Vyrastas commented on src/cli/cli.cpp:
```diff
@@ -5798,6 +5798,122 @@ template <> otError Interpreter::Process<Cmd("meshdiag")>(Arg aArgs[])
+     *     rx-on:no type:mtd full-net:no
+     *     rss - ave:-20 last:-20  margin:80
+     *     err-rate - frame:0.73% msg:0.00%
+     *     csl - sync:no period:0 timeout:0 channel:0
```
Need `Done` at the end?

Vyrastas commented on src/cli/cli.cpp:
```diff
@@ -5798,6 +5798,122 @@ template <> otError Interpreter::Process<Cmd("meshdiag")>(Arg aArgs[])
+     * child-rloc16: 0xdc03
+     *     fdde:ad00:beef:0:3aa6:b8bf:e7d6:eefe
+     *     fd00:2:0:0:8ff8:a188:7436:6720
+     *     fd00:1:0:0:1fcf:5495:790a:370f
```
Need `Done` at the end?

Vyrastas commented on src/cli/cli.cpp:
```diff
@@ -5798,6 +5798,122 @@ template <> otError Interpreter::Process<Cmd("meshdiag")>(Arg aArgs[])
+     * rloc16:0x7c00 ext-addr:4ed24fceec9bf6d3 ver:4
+     *    rx-on:yes type:ftd full-net:yes
+     *    rss - ave:-20 last:-20 margin:80
+     *    err-rate - frame:0.72% msg:0.00%
```
Need `Done` at the end?

Vyrastas commented on src/cli/cli.cpp:
```diff
@@ -5798,6 +5798,122 @@ template <> otError Interpreter::Process<Cmd("meshdiag")>(Arg aArgs[])
+     * * Device Mode
+     * * RSS (average and last)
+     * * Error rates: frame tx (at MAC layer), IPv6 message tx (above MAC)
+     * * CSL info
```
colon at end

CSL info:

abtink commented:
Thanks @Vyrastas. I added the suggestions above in https://github.com/openthread/openthread/pull/8866#issuecomment-1540973549 (which is the main PR adding "Child TLV and Neighbor TLV" for Network Diagnostic, etc).

This PR is for MLE Counters TLV (but I added it on top of the commit from the other PR to avoid conflict when we merge them). I may later include this in #8866 and close this PR.

abtink commented:
I include the commit in this PR into https://github.com/openthread/openthread/pull/8866. So we can close this.
